MoltenVK/MoltenVK/GPUObjects/MVKQueue.mm

```diff
@@ -312,7 +345,33 @@
 					device->getPhysicalDevice()->setConfigurationResult(VK_ERROR_DEVICE_LOST);
 					break;
 			}
+#if MVK_XCODE_12
+			if (mvkGetMVKConfiguration()->debugMode) {
+				if (&MTLCommandBufferEncoderInfoErrorKey != nullptr) {
+					if (NSArray<id<MTLCommandBufferEncoderInfo>>* mtlEncInfo = mtlCB.error.userInfo[MTLCommandBufferEncoderInfoErrorKey]) {
+						MVKLogInfo("Encoders for %p \"%s\":", mtlCB, mtlCB.label ? mtlCB.label.UTF8String : "");
+						for (id<MTLCommandBufferEncoderInfo> enc in mtlEncInfo) {
+							MVKLogInfo(" - %s: %s", enc.label.UTF8String, mvkStringFromErrorState(enc.errorState));
+							if (enc.debugSignposts.count > 0) {
+								MVKLogInfo("   Debug signposts:");
+								for (NSString* signpost in enc.debugSignposts) {
+									MVKLogInfo("    - %s", signpost.UTF8String);
+								}
+							}
+						}
+					}
+				}
+			}
+#endif
 		}
+#if MVK_XCODE_12
+		if (mvkGetMVKConfiguration()->debugMode) {
+			MVKLogInfo("Shader log messages:");
+			for (id<MTLFunctionLog> log in mtlCB.logs) {
+				MVKLogInfo("%s", log.description.UTF8String);
+			}
+		}
+#endif
```

Comment:
Hmmm...probably lots of under-the-covers Obj-C object creation going on as a result array copying and fast-enumeration. It's got me wondering if Metal makes sure to wrap callbacks in autorelease pools.

Couldn't find any docs on that.

In any case, since this is all debug activity, any leaks are unlikely to affect production, and its probably not worth the performance cost of wrapping every completion handler in an autorelease pool in production.